src/Avalonia.Controls.TreeDataGrid/Selection/TreeDataGridRowSelectionModel.cs

```diff
@@ -110,6 +113,99 @@ void ITreeDataGridSelectionInteraction.OnKeyDown(TreeDataGrid sender, KeyEventAr
+
+                string candidatePattern;
+                if (time < 5000000)
+                {
+                    if (_typedWord.Length == 1 && typedChar == _typedWord[0])
+                    {
+                        candidatePattern = _typedWord;
+                    }
+                    else
+                    {
+                        candidatePattern = _typedWord + typedChar;
+                    }
+                }
+                else
+                {
+                    candidatePattern = typedChar.ToString();
+                }
+                foreach (var column in treeDataGrid.Columns)
+                {
+                    if (column is ITextSearchable textSearchableColumn && textSearchableColumn.IsTextSearchEnabled == true)
+                    {
+                        Search(treeDataGrid, candidatePattern, selectedRowIndex, textSearchableColumn);
+                    }
+                    else if (column is HierarchicalExpanderColumn<TModel> hierarchicalColumn &&
+                        hierarchicalColumn.Inner is ITextSearchable textSearchableColumn2)
+                    {
+                        Search(treeDataGrid, candidatePattern, selectedRowIndex, textSearchableColumn2);
+                    }
+
+                }
+                _lastCharPressedTime = now;
+            }
+        }
+
+        private void Search(TreeDataGrid treeDataGrid, string candidatePattern, int selectedRowIndex, ITextSearchable column)
+        {
+            var textColumn = column as ISearchableTextColumn<TModel>;
+            var templateColumn = column as ISearchableTemplateColumn<TModel>;
+            var found = false;
+            var selector = textColumn == null ? templateColumn!.TextSearchValueSelector : textColumn.ResolveStingValueFromModel;
+            for (int i = candidatePattern.Length == 1 ? selectedRowIndex + 1 : selectedRowIndex; i <= _source.Rows.Count - 1; i++)
+            {
+                found = SearchAndSelectRow(treeDataGrid, candidatePattern, i, (TModel?)_source.Rows[i].Model, selector);
+                if (found)
+                {
+                    break;
+                }
+            }
+            if (!found)
+            {
+                for (int i = 0; i <= selectedRowIndex; i++)
+                {
+                    found = SearchAndSelectRow(treeDataGrid, candidatePattern, i, (TModel?)_source.Rows[i].Model, selector);
+                    if (found)
+                    {
+                        break;
+                    }
+                }
+            }
+        }
+
+        private bool SearchAndSelectRow(TreeDataGrid treeDataGrid,
+            string candidatePattern, int newIndex, TModel? model, Func<TModel, string?>? valueSelector)
+        {
+            if (valueSelector != null && model != null)
+            {
+                var value = valueSelector(model);
+                if (value != null && value.ToUpper().StartsWith(candidatePattern.ToUpper()))
```

Comment:
Might be worth storing the `candidatePattern` in uppercase to prevent having to call `ToUpper` on each row.